src/unravel/loop.cljs

```diff
@@ -311,9 +333,55 @@ interpreted by the REPL client. The following specials are available:
   [_ _ ctx]
   (interrupt ctx))
 
+(defn- line-up [rl]
+  (when-some [[_ prev-line curr-line] (re-find #"([^\r\n]*)(?:\r\n|\r|\n)([^\r\n]*)$" (subs (.-line rl) 0 (.-cursor rl)))]
+    (._moveCursor rl (- (inc (max (count prev-line) (count curr-line)))))))
+
+(defn- line-down [rl]
+  (when-some [[_ end-curr-line next-line] (re-find #"^([^\r\n]*)(?:\r\n|\r|\n)([^\r\n]*)" (subs (.-line rl) (.-cursor rl)))]
+    (let [start-curr-line (re-find #"[^\r\n]*$" (subs (.-line rl) 0 (.-cursor rl)))
+          curr-line (str start-curr-line end-curr-line)]
+      (._moveCursor rl (inc (min (+ (count end-curr-line) (count next-line)) (count curr-line)))))))
+
 (defmethod process[:readline :ready]
   [[_ rl] _ {:keys [sm connect] :as ctx}]
-  (let [ctx (assoc ctx :rl rl :completer-fn complete)]
+  (let [ctx (assoc ctx :rl rl :completer-fn complete)
+        send-input! (-> rl .-_line (.bind rl))
+        super-_ttyWrite (.-_ttyWrite rl)
+        super-_addHistory (.-_addHistory rl)
+        super-_historyPrev (.-_historyPrev rl)
+        super-_historyNext (.-_historyNext rl)]
+    (specify! rl
+      Object
+      (_line [this]
```

Comment:
It's not monkey patching, I don't change the prototype!
What I use seems to have been stable since I realized later that I was looking at some outdated code.
The deeper we go on rich editing, the more we'll rewrite of readline.

Reply:
Yeah maybe at some point we need to rewrite it entirely. For now this seems like a good solution